### DISCUSSION TITLE
Is the RetroBrew MF-PIC supported?

### DISCUSSION THREAD

**Discussion by sparkybrit** [wwarthen/RomWBW#294] [Q&A] [answered]
[https://www.retrobrewcomputers.org/doku.php?id=boards:ecb:mf-pic:start](url)

Is the RetroBrew MF-PIC supported?

**Comment by wwarthen** [comment, marked as answer]
Mostly... The UART, RTC, and Parallel/IDE interface are all fully supported.  The interrupt controller itself is not supported mostly due to lack of any specific need to do so.

Thanks,

Wayne

**Reply by sparkybrit** [reply]
Good enough! Thanks.

**Reply by sparkybrit** [reply]
It looks like the MFP support is off in the default MK4 config. 

Is this what I need to set in the MK4 config to turn MFP support on?
`DSRTCMODE .EQU DSRTCMODE_MFPIC`
`UARTMFP .EQU TRUE`
`PPIDE0BASE .EQU $10` (assuming $10 is the IO port jumper setting on the MFP)

**Comment by wwarthen** [comment]
The default base I/O address for the MFP board is $40.  If you use this address then, you would use the following in MK4_std.asm:

```
DSRTCMODE	.SET	DSRTCMODE_MFPIC	; DSRTC: OPERATING MODE: DSRTC_[STD|MFPIC]
UARTMFP		.SET	TRUE		; UART: AUTO-DETECT MF/PIC UART
PPIDEENABLE	.SET	TRUE		; PPIDE: ENABLE PARALLEL PORT IDE DISK DRIVER (PPIDE.ASM)
```

Note that I used `.SET` instead of `.EQU:`.  You must use .SET in the config files in the Config directory.  You should not be modifying the cfg_MK4.asm file in the HBIOS directory.

If  you want to change the base I/O address of the MFP board to $10, then you will need to modify the code in dsrtc.asm and ppide.asm to correspond to the correct I/O addresses.  The PPIDE driver can handle the alternate I/O port address by using `PPIDE0BASE`.

Thanks,

Wayne

**Reply by sparkybrit** [reply]
[Aren't](url) the Z180 internal io ports at $40?

`Z180_BASE	.EQU	$40		; Z180: I/O BASE ADDRESS FOR INTERNAL REGISTERS`

"new default is $10"
[MF-PIC 4.0 notes.pdf](https://github.com/wwarthen/RomWBW/files/9074353/MF-PIC.4.0.notes.pdf)

**Reply by wwarthen** [reply]
> [Aren't](url) the Z180 internal io ports at $40?
> 
> `Z180_BASE .EQU $40 ; Z180: I/O BASE ADDRESS FOR INTERNAL REGISTERS`
> 
> "new default is $10" [MF-PIC 4.0 notes.pdf](https://github.com/wwarthen/RomWBW/files/9074353/MF-PIC.4.0.notes.pdf)

Ugh, yes, you are absolutely correct.  The RomWBW code was created with the assumption that an interrupt controller is only needed for a Z80 since the Z180 has one built-in.  As you point out, John now suggests 0x10 as the default I/O base address, so I will change that in the official code base.

If you give me a few hours I will post code updates to the dev branch of RomWBW.  I won't be able to test the updates, but they are pretty simple.

Thanks,

Wayne

**Comment by wwarthen** [comment]
OK,  I just posted commit 3b0951f6d5e2568cba8832fbc26cd73c69f70782 with changes to accommodate placing the MFPIC at 0x10 instead of 0x40.  These changes are not tested, so please let me know if you have any problems.  The changes are pretty trivial, so they really should be OK.

Thanks,

Wayne

**Comment by sparkybrit** [comment]
Thanks. MK4_pic all working. This is the console output on the MF-PIC's UART

`RomWBW HBIOS v3.1.1-pre.179, 2022-07-11

MARK IV Z8S180-N @ 18.432MHz IO=0x40
0 MEM W/S, 2 I/O W/S, INT MODE 2, Z180 MMU
512KB ROM, 512KB RAM

ASCI0: IO=0x40 ASCI W/BRG MODE=38400,8,N,1
ASCI1: IO=0x41 ASCI W/BRG MODE=38400,8,N,1
UART0: IO=0x18 16550C MODE=38400,8,N,1 FIFO AFC
DSRTC: MODE=MFPIC IO=0x13 Sat 2000-01-01 00:00:02 CHARGE=OFF
CVDU: MODE=ECB EGA IO=0xE4 NOT PRESENT
VGA: IO=0xE2 NOT PRESENT
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=DIDE IO=0x2A NOT PRESENT
IDE: IO=0x80 MODE=MK4
IDE0: NO MEDIA
IDE1: NO MEDIA
PPIDE: IO=0x14
PPIDE0: LBA BLOCKS=0x00EEC9B0 SIZE=7641MB
PPIDE1: NO MEDIA
PPIDE: IO=0x20 PPI NOT PRESENT
SD: MODE=MK4 FAST OPR=0x89 CNTR=0x4A TRDR=0x4B DEVICES=1
SD0: SDHC NAME=SS32G BLOCKS=0x03B72400 SIZE=30436MB
PRP: IO=0xA8 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      ASCI0:      RS-232            38400,8,N,1
Char 1      ASCI1:      RS-232            38400,8,N,1
Char 2      UART0:      RS-232            38400,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       Hard Disk         --
Disk 3      IDE1:       Hard Disk         --
Disk 4      PPIDE0:     CompactFlash      7641MB,LBA
Disk 5      PPIDE1:     Hard Disk         --
Disk 6      SD0:        SD Card           30436MB,LBA


MARK IV Boot Loader

Boot [H=Help]:`

**Reply by wwarthen** [reply]
Excellent.  I was pretty sure it would, but have never really put together this specific combination of hardware.

I would expect that the MF/PIC handles a wider range of CF Cards via it's PPIDE interface.

Thanks,

Wayne
